[MRG+1] MAINT Replace manual checks with `check_is_fitted`

agamemnonc opened the pull request:
#### Reference Issues/PRs
Fixes #12991.

#### What does this implement/fix? Explain your changes.
Replaces manual checks with `check_is_fitted` utility function in various places. 

#### Any other comments?
All modified files have been checked with flake8 and autopep8 and any formatting issues have been addressed.

agamemnonc committed:
online_lda and tests

agamemnonc committed:
forest

agamemnonc committed:
linear model base

agamemnonc committed:
logistic

agamemnonc committed:
Merge branch 'master' into check_is_fitted_replacements

agamemnonc committed:
birch

agamemnonc committed:
pep8 fix

adrinjalali commented on sklearn/decomposition/online_lda.py:
```diff
@@ -379,7 +378,7 @@ def _e_step(self, X, cal_sstats, random_init, parallel=None):
         if parallel is None:
             parallel = Parallel(n_jobs=n_jobs, verbose=max(0,
-                                self.verbose - 1))
+                                                           self.verbose - 1))
```
probably break the line at `verbose`.

agamemnonc replied:
Thanks for reviewing, now fixed.

adrinjalali commented:
Although there are quite some changes that are PEP8 related and not directly related to this PR, LGTM, if tests pass.

agamemnonc committed:
fix formatting issue

agamemnonc committed:
exceptions doc example

agamemnonc commented:
> Although there are quite some changes that are PEP8 related and not directly related to this PR, LGTM, if tests pass.

Thanks. Indeed, as per the description above, I addressed all formatting issues in the modified files so that autopep8/flake8 checks passed before submitting the PR.

jnothman commented on sklearn/ensemble/forest.py:
```diff
@@ -990,6 +989,7 @@ class labels (multi-output problem).
     --------
     DecisionTreeClassifier, ExtraTreesClassifier
     """
+
```
I don't think this is a convention we adhere to, even though I see pep257 asks for it.

jnothman replied:
I think the """ leaves enough space

jnothman commented on sklearn/exceptions.py:
```diff
@@ -30,7 +30,8 @@ class NotFittedError(ValueError, AttributeError):
     ...     print(repr(e))
     ...                        # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
-    NotFittedError('This LinearSVC instance is not fitted yet'...)
+    NotFittedError("This LinearSVC instance is not fitted yet. Call 'fit' with
```
Not sure why you changed this

agamemnonc replied:
I had to change this, otherwise tests would not pass (link to previously failed test [here](https://travis-ci.org/scikit-learn/scikit-learn/jobs/481309007#L2125)), the reason being that the previous manual check had its own error message which was not consistent with the default 'check_is_fitted' error message.

jnothman replied:
Of course. Thanks for the response.

jnothman reviewed (commented):
The core parts of this PR seem okay, when I can find them

agamemnonc commented:
> The core parts of this PR seem okay, when I can find them

Thanks for reviewing,

OK, I could revert the formatting changes if that would be preferred (your first two comments above)?
Regarding the modification in `exceptions.py` please see my response above.

glemaitre commented:
@agamemnonc Could you revert the style changes. I can make a review then.

agamemnonc committed:
revert style changes

agamemnonc committed:
one more style revert

agamemnonc commented:
> @agamemnonc Could you revert the style changes. I can make a review then.

OK, done; please review.

Of course, there are now some flake8 warnings on the modified files (mostly due to long lines).

glemaitre commented:
I think that we have some missing occurrences:

https://github.com/scikit-learn/scikit-learn/blob/fdf2f3834e4772b401728ef8ebd280792dfdaf65/sklearn/linear_model/base.py#L253-L255

https://github.com/scikit-learn/scikit-learn/blob/fdf2f3834e4772b401728ef8ebd280792dfdaf65/sklearn/linear_model/logistic.py#L1645-L1646

I also think that we should change the common test:

https://github.com/scikit-learn/scikit-learn/blob/fdf2f3834e4772b401728ef8ebd280792dfdaf65/sklearn/utils/estimator_checks.py#L1586-L1616

with something like:

```python
@ignore_warnings
def check_estimators_unfitted(name, estimator_orig):
    """Check that predict raises an exception in an unfitted estimator.

    Unfitted estimators should raise either AttributeError or ValueError.
    The specific exception type NotFittedError inherits from both and can
    therefore be adequately raised for that purpose.
    """

    # Common test for Regressors, Classifiers and Outlier detection estimators
    X, y = _boston_subset()

    estimator = clone(estimator_orig)

    msg = ("{} instance is not fitted yet. Call 'fit' with appropriate "
           "arguments".format(estimator.__class__.__name__))
    for method in ('decision_function', 'predict', 'predict_proba',
                   'predict_log_proba'):
        if getattr(estimator, method, None) is not None:
            assert_raises_regex(NotFittedError, msg,
                                getattr(estimator, method), X)
```

glemaitre commented:
@jnothman Do you know why we were checking `AttributeError` and `ValueError` instead of directly `NotFittedError`?

glemaitre commented on sklearn/cluster/birch.py:
```diff
@@ -537,16 +537,11 @@ def partial_fit(self, X=None, y=None):
-            raise NotFittedError("Fit training data before predicting")
-
-        if is_fitted and X.shape[1] != self.subcluster_centers_.shape[1]:
+        if hasattr(self, 'subcluster_centers_') and X.shape[1] != \
```
We already know that it has `subcluster_centers_` otherwise the `check_is_fitted` would have already raised an error.

```python
if X.shape[1] != self.subcluster_centers_.shape[1]:
    ...
```

agamemnonc replied:
The `check_is_fitted` is called with two arguments `'subcluster_centers_` and `partial_fit_` and `all_or_any` is set to `any`, so it is not always the case that the object has `subcluster_centers_ `. To convince yourself, remove the `hasattr(self, 'subcluster_centers_')` and you will see that one of the tests fails.

glemaitre replied:
Oh ok, I missed that. Could you then remove the `\` and use parentheses instead.

glemaitre commented on sklearn/decomposition/online_lda.py:
```diff
@@ -19,9 +19,8 @@
                      gen_batches, gen_even_slices)
 from ..utils.fixes import logsumexp
-from ..utils.validation import check_non_negative
+from ..utils.validation import check_non_negative, check_is_fitted
```
```suggestion
from ..utils.validation import check_non_negative
from ..utils.validation import check_is_fitted
```

glemaitre replied:
We don't have a real convention for that.

My personal thoughts on that are the following:

* import that we are almost sure to not modify can go on a single line (this is the case with joblib)
* imports that are more probable to be changed (this is the case with validation --- we might add or remove an import in the future) could go on different lines. On a single line, we can get more diff just for reordering import and line breaking.

But I agree that this is only my personal taste. Both are valid and I will not block the merge for that. Choose what you prefer.

agamemnonc replied:
Agreed -- will change.

glemaitre commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -184,7 +184,8 @@ def test_lda_no_component_error():
     X = rng.randint(4, size=(20, 10))
     lda = LatentDirichletAllocation()
-    regex = r"^no 'components_' attribute"
+    regex = "This LatentDirichletAllocation instance is not fitted yet. " + \
```
We do not use `\` to break line most of the time but the parentheses instead.

```python
regex = ("This LatentDirichletAllocation instance is not fitted yet. "
         "Call 'fit' with appropriate arguments before using this method.")

```

agamemnonc replied:
Fixed

glemaitre commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -184,7 +184,8 @@ def test_lda_no_component_error():
-    regex = r"^no 'components_' attribute"
+    regex = "This LatentDirichletAllocation instance is not fitted yet. " + \
+        "Call 'fit' with appropriate arguments before using this method."
     assert_raises_regexp(NotFittedError, regex, lda.transform, X)
```
This already tested by the common test:

```
sklearn/tests/test_common.py::test_non_meta_estimators[LatentDirichletAllocation-LatentDirichletAllocation-check_transformers_unfitted]
```

We can remove it then

agamemnonc replied:
I am afraid I am not sure where exactly this is tested. Could you please provide a permalink from the code so that I can see it more clearly?

glemaitre replied:
This is `estimator_check.py` This is the common test that you changed below. It is run for most of the estimator and the line that I paste is the relevant run for this test for LDA.

glemaitre commented on sklearn/exceptions.py:
```diff
@@ -30,7 +30,8 @@ class NotFittedError(ValueError, AttributeError):
     ...                        # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
-    NotFittedError('This LinearSVC instance is not fitted yet'...)
+    NotFittedError("This LinearSVC instance is not fitted yet. Call 'fit' with
+    appropriate arguments before using this method."...)
```
You don't need the ellipsis in this case.

```suggestion
    appropriate arguments before using this method.")
```

glemaitre commented on sklearn/exceptions.py:
```diff
@@ -30,7 +30,8 @@ class NotFittedError(ValueError, AttributeError):
     ... except NotFittedError as e:
     ...     print(repr(e))
     ...                        # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
```
```suggestion
    ...                        # doctest: +NORMALIZE_WHITESPACE
```

glemaitre commented on sklearn/linear_model/logistic.py:
```diff
@@ -28,10 +28,9 @@
 from ..utils.fixes import logsumexp
 from ..utils.optimize import newton_cg
-from ..utils.validation import check_X_y
+from ..utils.validation import check_X_y, check_is_fitted
```
```suggestion
from ..utils.validation import check_X_y
from ..utils.validation import check_is_fitted
```

agamemnonc replied:
Please see my response above.

glemaitre replied:
as you wish :)

glemaitre reviewed (changes requested):
I will check if we don't have redundant tests but you can already address those comments.

glemaitre commented:
Regarding the tests, I would propose to change the following:

https://github.com/scikit-learn/scikit-learn/blob/fdf2f3834e4772b401728ef8ebd280792dfdaf65/sklearn/ensemble/tests/test_forest.py#L371-L378

```python
@pytest.mark.parametrize('name', FOREST_ESTIMATORS)
def test_unfitted_feature_importances(name):
    err_msg = ('This {} instance is not fitted yet. Call 'fit' with appropriate '
               'arguments before using this method.'.format(name))
    pytest.raises(NotFittedError, match=err_msg):
         gettattr(FOREST_ESTIMATORS[name](), 'feature_importances')
```

glemaitre commented:
Even if we don't touch a public API, I would add an entry in the what's new since we modified the tests and error message.

agamemnonc committed:
update utils/estimator_checks and tests

agamemnonc committed:
style fix

agamemnonc committed:
Merge branch 'master' into check_is_fitted_replacements

agamemnonc committed:
remove ellipsis

agamemnonc commented:
> I think that we have some missing occurrences:
> EDIT: sorry I did not checkout the good PR
> 
> I also think that we should change the common test:
> 
> [scikit-learn/sklearn/utils/estimator_checks.py](https://github.com/scikit-learn/scikit-learn/blob/fdf2f3834e4772b401728ef8ebd280792dfdaf65/sklearn/utils/estimator_checks.py#L1586-L1616)
> 
> Lines 1586 to 1616 in [fdf2f38](/scikit-learn/scikit-learn/commit/fdf2f3834e4772b401728ef8ebd280792dfdaf65)
> 
>  def check_estimators_unfitted(name, estimator_orig): 
>      """Check that predict raises an exception in an unfitted estimator. 
>   
>      Unfitted estimators should raise either AttributeError or ValueError. 
>      The specific exception type NotFittedError inherits from both and can 
>      therefore be adequately raised for that purpose. 
>      """ 
>   
>      # Common test for Regressors, Classifiers and Outlier detection estimators 
>      X, y = _boston_subset() 
>   
>      estimator = clone(estimator_orig) 
>   
>      msg = "fit" 
>   
>      if hasattr(estimator, 'predict'): 
>          assert_raise_message((AttributeError, ValueError), msg, 
>                               estimator.predict, X) 
>   
>      if hasattr(estimator, 'decision_function'): 
>          assert_raise_message((AttributeError, ValueError), msg, 
>                               estimator.decision_function, X) 
>   
>      if hasattr(estimator, 'predict_proba'): 
>          assert_raise_message((AttributeError, ValueError), msg, 
>                               estimator.predict_proba, X) 
>   
>      if hasattr(estimator, 'predict_log_proba'): 
>          assert_raise_message((AttributeError, ValueError), msg, 
>                               estimator.predict_log_proba, X) 
>   
> with something like:
> 
> ```python
> @ignore_warnings
> def check_estimators_unfitted(name, estimator_orig):
>     """Check that predict raises an exception in an unfitted estimator.
> 
>     Unfitted estimators should raise a NotFittedError.
>     """
> 
>     # Common test for Regressors, Classifiers and Outlier detection estimators
>     X, y = _boston_subset()
> 
>     estimator = clone(estimator_orig)
> 
>     msg = ("{} instance is not fitted yet. Call 'fit' with appropriate "
>            "arguments".format(estimator.__class__.__name__))
>     for method in ('decision_function', 'predict', 'predict_proba',
>                    'predict_log_proba'):
>         if getattr(estimator, method, None) is not None:
>             assert_raises_regex(NotFittedError, msg,
>                                 getattr(estimator, method), X)
> ```

This is now fixed. thanks!

agamemnonc committed:
style fix

agamemnonc committed:
style fix

agamemnonc committed:
another style fix

agamemnonc committed:
test_unfitted_feature_importances update

agamemnonc commented:
> test_unfitted_feature_importances

Thanks, updated.

agamemnonc committed:
update whats_new rst file

agamemnonc commented:
> Even if we don't touch a public API, I would add an entry in the what's new since we modified the tests and error message.

Sure, done.

agamemnonc committed:
another style fix

agamemnonc committed:
remove redundant lda estimator check

agamemnonc commented:
@glemaitre many thanks for the thorough review! 

I think I have now addressed all your comments --please have a look again let me know if there is anything else.

agamemnonc committed:
revert previous commit and remove lda.transofrm check only

agamemnonc commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -179,15 +179,6 @@ def test_lda_negative_input():
-    lda = LatentDirichletAllocation()
-    regex = r"^no 'components_' attribute"
-    assert_raises_regexp(NotFittedError, regex, lda.transform, X)
-    assert_raises_regexp(NotFittedError, regex, lda.perplexity, X)
```
Of course, you are right. 

This is now fixed, sorry about that.

glemaitre committed:
nitpicks

glemaitre committed:
DOC: add missing backsticks

glemaitre reviewed (approved):
I made 2 small push for some glitches. LGTM.
@adrinjalali could you have another look and merge if you are fine with the changes.

glemaitre commented:
@jnothman also add a first look

agamemnonc committed:
comment typo fix

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1586,33 +1587,22 @@ def check_estimators_fit_returns_self(name, estimator_orig,
     estimator = clone(estimator_orig)
 
-    msg = "fit"
+    msg = ("{} instance is not fitted yet. Call 'fit' with appropriate "
```
Making this requirement more strict also places the same constraint on developers who use check_estimator. For other kinds of backwards compatibility issues we would raise a deprecation warning and still pass the test, rather than raise an error outright. I would consider just strengthening this test if we don't think developers will be upset about having their compatibility broken all of a sudden. In any case, this change needs to be separately documented under "Changes to estimator checks"

glemaitre replied:
On the one hand, I think that you are completely right and it would not cost that much to revert to the previous match (@agamemnonc could you revert it).

On the other hand, it happens in the past that scikit-learn broke some of the backward compatibility in the common test and I personally did not mind because it was always toward more consistent checks. For instance, if the estimator tags come to break some of the estimator checks, I would think this for the best since it would allow more modularity.

agamemnonc replied:
Please could you clarify whether you would like me to revert to previous state or include the new error message with a deprecation warning for the old one.

Also some leads re: the rst file "Changes to estimator checks" update would be appreciated, thanks.

jnothman reviewed (commented):
As a third party package developer, your opinion on backwards compatibility of estimator checks is very welcome @glemaitre

glemaitre commented:
> As a third party package developer, your opinion on backwards compatibility of estimator checks is very welcome @glemaitre

I personally don't mind the change, but our third-party package is quite limited in size and it is easy to make changes related to integration with scikit-learn. However, I might think that it could be more problematic for some larger packages. So I would say that until we can keep backward compatibility, we should do it. Breaking the backwards-compatibility for the checks might still be useful when we make an improvement which we considered beneficial for the ecosystem.

agamemnonc commented on doc/whats_new/v0.21.rst:
```diff
@@ -286,3 +286,6 @@ These changes mostly affect library developers.
   `predict`, `predict_proba`, `transform`, and `decision_function` does not
   change. :issue:`12328` by :user:`Nicolas Hug <NicolasHug>`
+
+- Replace manual checks with ``check_is_fitted`` and ensure that error messages
```
I am not sure I am following this, did you perhaps mean
"Not fitted error messages are now more uniform." ?

jnothman commented:
We can satisfy all needs by deprecating "until 0.20, predicting without
fitting should have raised either a ValueError or AttributeError. From 0.23
a NotFittedError with the following format of error message is required..."
and raise a warning during the deprecation period.

Regarding changes to estimator checks, see the similar section in v0.20.rst

agamemnonc committed:
update whats_new rst file

agamemnonc commented:
I am not too sure how to include the deprecation in ``check_estimators_unfitted``. 

I have tried the following but doesn't pass the tests, any help @jnothman @glemaitre would be appreciated. 

```python
@ignore_warnings
def check_estimators_unfitted(name, estimator_orig):
    """Check that predict raises an exception in an unfitted estimator.

    Unfitted estimators should raise a NotFittedError.
    """

    # Common test for Regressors, Classifiers and Outlier detection estimators
    X, y = _boston_subset()

    estimator = clone(estimator_orig)

    msg = ("{} instance is not fitted yet. Call 'fit' with appropriate "
           "arguments".format(estimator.__class__.__name__))

    for method in ('decision_function', 'predict', 'predict_proba',
                   'predict_log_proba'):
        if getattr(estimator, method, None) is not None:
            try:
                assert_raises_regex(NotFittedError, msg,
                                    getattr(estimator, method), X)
            except AssertionError:
                msg = "fit"
                assert_raises_regex((AttributeError, ValueError), msg,
                                    getattr(estimator, method), X)
                deprecated(("Until 0.20, predicting without fitting "
                           "should have raised either a ValueError or "
                           "AttributeError. From 0.23 a NotFittedError "
                           "with the following format of error message "
                           "is required: <class_name> instance is not fitted "
                           "yet. Call 'fit' with appropriate arguments."))
```

Also, please review the changes in v0.21.rst and let me know if you think they look OK.

Thanks!

jnothman commented:
Brief for now: Using pytest.raises might make the code much simpler than
using assert_raises_*

glemaitre commented:
You can use `getattr(estimator, method, None) is not None:` to get the default to None.
It avoids the `try ... except`

agamemnonc commented:
Yes, that's what the code currently looks like:
```python
msg = ("{} instance is not fitted yet. Call 'fit' with appropriate "
       "arguments".format(estimator.__class__.__name__))

for method in ('decision_function', 'predict', 'predict_proba',
               'predict_log_proba'):
    if getattr(estimator, method, None) is not None:
        assert_raises_regex(NotFittedError, msg,
                            getattr(estimator, method), X)
```
The problem is I am not too sure how to include the deprecation here, i.e. cover the case where an Attribute or ValueError is raised instead with the previous error message ("fit"), in order to allow that and issue a DeprecationWarning.

glemaitre commented:
Oh I see why you wanted a try except then. Basically use `pytest.raises` and check https://docs.pytest.org/en/latest/assert.html

I think something around:

```python
with pytest.raises(NotFittedError) as excinfo:
    getattr(estimator, method), X)

if not str(excinfo.value) in msg and 'fit' in str.excinfo.value):
    # raise deprecation waring
else:
    assert 'fit' in str.excinfo.value
```

I wrote this pretty quickly. That might be buggy

jnothman commented:
Please resolve conflicts with master.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1586,33 +1587,22 @@ def check_estimators_fit_returns_self(name, estimator_orig,
 
-    msg = "fit"
+    msg = ("{} instance is not fitted yet. Call 'fit' with appropriate "
+           "arguments".format(estimator.__class__.__name__))
```
I don't know whether this is too hard a constraint, i.e. whether there are cases you don't just want to say the class name.

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -286,3 +294,9 @@ These changes mostly affect library developers.
   `predict`, `predict_proba`, `transform`, and `decision_function` does not
   change. :issue:`12328` by :user:`Nicolas Hug <NicolasHug>`
+
+- ``check_estimators_unfitted`` should now raise a ``NotFittedError`` for an
```
This is not quite right. "Estimators are now expected to raise a NotFittedError if `predict` or `transform` is called before `fit`; previously an AttributeError or ValueError was acceptable".

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -286,3 +294,9 @@ These changes mostly affect library developers.
+
+- ``check_estimators_unfitted`` should now raise a ``NotFittedError`` for an
+  unfitted estimator. Previously raised ``AttributeError`` and ``ValueError``
+  have been deprecated. The old error message "fit" has been deprecated in favor
```
I can't see anything deprecated. I can see things changed without deprecation.

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -286,3 +294,9 @@ These changes mostly affect library developers.
+- ``check_estimators_unfitted`` should now raise a ``NotFittedError`` for an
+  unfitted estimator. Previously raised ``AttributeError`` and ``ValueError``
+  have been deprecated. The old error message "fit" has been deprecated in favor
+  of "<class_name> instance is not fitted yet. Call 'fit' with appropriate arguments".
```
Note that they should use `check_is_fitted`

jnothman reviewed (commented):
I appreciate the consistent use of `check_is_fitted` within the library, but I'm not entirely sure we should be forcing all library developers to use the exact same error message. The default message does not, for instance, mention `partial_fit`.

jnothman commented:
When you get around to resolving my comments, please also move your change log entry to v0.22.rst as version 0.21 has been released.

glemaitre committed:
Merge remote-tracking branch 'origin/master' into pr/agamemnonc/13013

glemaitre committed:
reviews

glemaitre committed:
add tag

glemaitre commented:
@jnothman GaussianProcessRegressor can work without calling `fit`. I added a tag `requires_fit` but I am not sure if it would be something that we want.

WDYT?

jnothman commented:
I think we have explicitly tried to support GPs without fit. Otherwise you
upset all the bayesians.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1650,47 +1653,21 @@ def check_estimators_fit_returns_self(name, estimator_orig,
-    if hasattr(estimator, 'predict_log_proba'):
-        assert_raise_message((AttributeError, ValueError), msg,
-                             estimator.predict_log_proba, X)
+    msg = "not fitted yet. Call 'fit' with appropriate arguments"
```
What do you think of my critique of requiring this message? I'd be happy to require NotFittedError without constraining the message

glemaitre replied:
I am fine with it as well. Enforcing the message is maybe not necessary by the end.

glemaitre commented:
> I think we have explicitly tried to support GPs without fit. Otherwise you upset all the bayesians.

And do you think that having the tag `requires_fit` (default to True) is a good idea?

glemaitre committed:
be more permissive and document tag

glemaitre commented:
@jnothman could you have another look. Apart of the tag I think that the PR is good to be merged.

glemaitre committed:
Merge remote-tracking branch 'origin/master' into pr/agamemnonc/13013

NicolasHug commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1650,47 +1653,16 @@ def check_estimators_fit_returns_self(name, estimator_orig,
-                             estimator.predict_log_proba, X)
+    for method in ('decision_function', 'predict', 'predict_proba',
+                   'predict_log_proba'):
+        if getattr(estimator, method, None) is not None:
```
use `hasattr` instead?

NicolasHug commented on doc/whats_new/v0.22.rst:
```diff
@@ -48,7 +48,19 @@ Changelog
+Miscellaneous
+.............
+
+- |API| Replace manual checks with ``check_is_fitted``. Not fitted errors are
```
May I suggest

> Errors thrown when using a non-fitted estimators are now more uniform

NicolasHug commented on doc/developers/contributing.rst:
```diff
@@ -1512,6 +1512,10 @@ stateless
     an estimator is stateless, it might still need a call to ``fit`` for initialization.
 
+requires_fit
+    whether the estimator requires to be fitted before to call one of
```
```suggestion
    whether the estimator requires to be fitted before calling one of
```

NicolasHug commented on doc/developers/contributing.rst:
```diff
@@ -1512,6 +1512,10 @@ stateless
 
+requires_fit
+    whether the estimator requires to be fitted before to call one of
+    ``transform``, ``predict``, ``predict_proba``, or ``decision_function``.
```
The method names (here and below) could be using single backticks to enable links to the glossary

NicolasHug reviewed (commented):
A few niticks.

I think the introduction of the estimator tag is appropriate here.

I'm not pressing "Approve" because TBH I'm not entirely sure what should be done regarding the deprecation, but I'm tending towards LGTM.

agamemnonc committed:
replace getattr with hasattr

agamemnonc committed:
fix rst file entry

agamemnonc committed:
fix typo in developers/contributing

agamemnonc committed:
developers/contributing links to glossary

agamemnonc commented:
Thanks @NicolasHug for reviewing. 

I think I have now addressed the issues you raised, otherwise please let me know.

jnothman reviewed (approved):
Thanks @agamemnonc!

glemaitre commented:
Thanks @agamemnonc

agamemnonc commented:
> Thanks @agamemnonc

Thank you @glemaitre  for your input and everyone else for reviewing.
